typescript/src/backoff.ts

```diff
@@ -0,0 +1,97 @@
+type ErrorMatcherFn = (err: unknown) => boolean
+
+/**
+ * @callback RetrierFn A function that can retry any function passed to it a
+ *           set number of times, returning its result when it succeeds. Created
+ *           using `backoffRetrier`.
+ * @param {function(): Promise<T>} fn The function to be retried.
+ * @return {Promise<T>}
+ */
+
+/**
+ * A function that is called with execution status messages.
+ */
+type ExecutionLoggerFn = (msg: string) => void
+
+/**
+ * Returns a retrier that can be passed a function to be retried `retries`
+ * number of times, with exponential backoff. The result will return the
+ * function's return value if no exceptions are thrown. It will only retry if
+ * the function throws an exception matched by `matcher`; {@see retryAll } can
+ * be used to retry no matter the exception, though this is not necessarily
+ * recommended in production.
+ *
+ * Example usage:
+ *
+ *      await url.get("https://example.com/") // may transiently fail
+ *      // Retries 3 times with exponential backoff, no matter what error is
+ *      // reported by `url.get`.
+ *      backoffRetrier(3)(async () => url.get("https://example.com"))
+ *      // Retries 3 times with exponential backoff, but only if the error
+ *      // message includes "server unavailable".
+ *      backoffRetrier(3, (_) => _.message.includes('server unavailable'))(
+ *        async () => url.get("https://example.com"))
+ *      )
+ *
+ * @param {number} retries The number of retries to perform before bubbling the
+ *        failure out.
+ * @param {ExecutionLoggerFn} logger A logger function to pass execution messages.
+ * @param {ErrorMatcherFn} [errorMatcher=retryAll] A matcher function that
+ *        receives the error when an exception is thrown, and returns true if
+ *        the error should lead to a retry. A false return will rethrow the
+ *        error and terminate the retry loop.
+ * @returns {RetrierFn<T>} A function that can retry any function.
+ */
+export function backoffRetrier<T>(
+  retries: number,
+  logger: ExecutionLoggerFn = console.debug,
+  errorMatcher: ErrorMatcherFn = retryAll
+) {
+  return async (fn: () => Promise<T>): Promise<T> => {
+    for (let attempt = 0; attempt < retries; attempt++) {
+      try {
+        logger(`making attempt number ${attempt}`)
+
+        return await fn()
+      } catch (error: unknown) {
+        if (!errorMatcher(error)) {
+          // If the matcher doesn't match this error, rethrow and stop.
+          throw error
+        }
+
+        const backoffMillis = Math.pow(2, attempt) * 1000
+        const jitterMillis = Math.floor(Math.random() * 100)
+        const waitMillis = backoffMillis + jitterMillis
+
+        logger(
+          `attempt ${attempt} failed: ${error}; ` +
+            `retrying after ${waitMillis} milliseconds`
+        )
+
+        await new Promise((resolve) => setTimeout(resolve, waitMillis))
+      }
+    }
+
+    // Last attempt, unguarded.
+    return await fn()
```

Comment:
Is there any sense to `await` here? We need to return a promise anyway. Also, the ESLint complains: "Returning an awaited promise is not allowed in this context."

Reply:
Niche catch. I updates it as part of https://github.com/keep-network/tbtc-v2/pull/516/commits/ccd4d0a9b2a6be5392151cd5c7dee259ef89f69c#diff-6f0abbdd421b0e7a7890b8e6ede7a16fb44fc7dd491c59dcd21e9888583d2476R117